app/wasm/message_plugin.go

```diff
@@ -0,0 +1,201 @@
+
+func (m *CustomMessenger) mintTokens(ctx sdk.Context, contractAddr sdk.AccAddress, mint *wasmbindings.MintTokens) ([]sdk.Event, [][]byte, error) {
+	err := PerformMint(m.bank, ctx, contractAddr, mint)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "perform mint")
+	}
+	return nil, nil, nil
+}
+
+func PerformMint(b *bankkeeper.BaseKeeper, ctx sdk.Context, contractAddr sdk.AccAddress, mint *wasmbindings.MintTokens) error {
+	if mint == nil {
+		return wasmvmtypes.InvalidRequest{Err: "mint token null mint"}
+	}
+	rcpt, err := parseAddress(mint.Recipient)
+	if err != nil {
+		return err
+	}
+
+	denom, err := GetFullDenom(contractAddr.String(), mint.SubDenom)
+	if err != nil {
+		return sdkerrors.Wrap(err, "mint token denom")
+	}
+	if mint.Amount.IsNegative() {
+		return wasmvmtypes.InvalidRequest{Err: "mint token negative amount"}
+	}
+	coins := []sdk.Coin{sdk.NewCoin(denom, mint.Amount)}
+
+	err = b.MintCoins(ctx, gammtypes.ModuleName, coins)
+	if err != nil {
+		return sdkerrors.Wrap(err, "minting coins from message")
+	}
+	err = b.SendCoinsFromModuleToAccount(ctx, gammtypes.ModuleName, rcpt, coins)
+	if err != nil {
+		return sdkerrors.Wrap(err, "sending newly minted coins from message")
+	}
+	return nil
+}
+
+func (m *CustomMessenger) swapTokens(ctx sdk.Context, contractAddr sdk.AccAddress, swap *wasmbindings.SwapMsg) ([]sdk.Event, [][]byte, error) {
+	_, err := PerformSwap(m.gammKeeper, ctx, contractAddr, swap)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "perform swap")
+	}
+	return nil, nil, nil
+}
+
+// PerformSwap can be used both for the real swap, and the EstimateSwap query
+func PerformSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.AccAddress, swap *wasmbindings.SwapMsg) (*wasmbindings.SwapAmount, error) {
+	if swap == nil {
+		return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap null swap"}
+	}
+	if swap.Amount.ExactIn != nil {
+		routes := []gammtypes.SwapAmountInRoute{{
+			PoolId:        swap.First.PoolId,
+			TokenOutDenom: swap.First.DenomOut,
+		}}
+		for _, step := range swap.Route {
+			routes = append(routes, gammtypes.SwapAmountInRoute{
+				PoolId:        step.PoolId,
+				TokenOutDenom: step.DenomOut,
+			})
+		}
+		if swap.Amount.ExactIn.Input.IsNegative() {
+			return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap negative amount in"}
+		}
+		tokenIn := sdk.Coin{
+			Denom:  swap.First.DenomIn,
+			Amount: swap.Amount.ExactIn.Input,
+		}
+		tokenOutMinAmount := swap.Amount.ExactIn.MinOutput
+		estimatedAmount, err := keeper.MultihopSwapExactAmountIn(ctx, contractAddr, routes, tokenIn, tokenOutMinAmount)
+		if err != nil {
+			return nil, sdkerrors.Wrap(err, "gamm perform swap exact amount in")
+		}
+		return &wasmbindings.SwapAmount{Out: &estimatedAmount}, nil
+	} else if swap.Amount.ExactOut != nil {
+		routes := []gammtypes.SwapAmountOutRoute{{
+			PoolId:       swap.First.PoolId,
+			TokenInDenom: swap.First.DenomIn,
+		}}
+		output := swap.First.DenomOut
```

Comment:
Note that we simplify the Rust API such that it always flows from first to last (even if we set ExactOut), and then convert to the different Go API here. This shows how we can easily adapt to new APIs as long as the same basic functionality is exposed.